pkg/gardenlet/controller/federatedseed/shootstate/common_sync.go

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shootstate
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	apiextensions "github.com/gardener/gardener/pkg/api/extensions"
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	gardencorev1alpha1helper "github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func createEnqueueOnAddFunc(queue workqueue.RateLimitingInterface) func(extensionObject interface{}) {
+	return func(newObj interface{}) {
+		extensionObject := newObj.(extensionsv1alpha1.Object)
+		if extensionObject.GetExtensionStatus().GetState() == nil {
+			return
+		}
+		enqueue(queue, newObj)
+	}
+}
+
+func createEnqueueOnUpdateFunc(queue workqueue.RateLimitingInterface) func(newExtensionObject, oldExtensionObject interface{}) {
+	return func(newObj, oldObj interface{}) {
+		var (
+			newExtensionObj = newObj.(extensionsv1alpha1.Object)
+			oldExtensionObj = oldObj.(extensionsv1alpha1.Object)
+		)
+
+		if newExtensionObj.GetExtensionStatus().GetState() == nil {
+			return
```

Comment:
Hm, should we really return here? What if an extension controller removed state from its resource (because it's no longer needed)? In this case we should also remove it from the `ShootState` to not confuse it during a migration.

Reply:
ping

Reply:
With the current implementation, if the state of the extension resource is  removed we are removing it from the ShootState.